### PR TITLE
ndarray: more Base-like APIs

### PR DESCRIPTION
**Pull request by iblis17**
- remapping
    - `mean`
    - `sum`
    - `maximum`
    - `minimum`
    - `permutedims`
    - `prod`

- also fix the `axis` value mapping
- `mean(arr, axis=0)` is not Julian

### CODE REVIEW

**Comment by pluskid** [comment]
min, max, etc?

**Comment by iblis17** [comment]
Done for min and max. Anything else I missed?

**Comment by pluskid** [comment]
I'm searching the doc for the keyword 'dim' or 'axis', it seems there are a lot of operators to consider (e.g. `prod`). I'm wondering if there is better way to handle this. E.g. detecting in the meta data from libmxnet whether that argument is called `dim` / `axis`?

**Comment by iblis17** [comment]
> I'm wondering if there is better way to handle this. E.g. detecting in the meta data from libmxnet whether that argument is called `dim` / `axis`?

I guess that docstring is the only set we can search. doesnt see any API about the function signature or similar.

I still prefer to do it manually.
I think a better way will be:
1. add a utility script which does some text seaching/parsing for function signature
2. run it in CI (maybe) and it pops up some warning
3. then, (if someone find these warning message) open an issue for remapping the functions that bug by this script.
4. one of the purpose of manually remapping function is add test cases for them, in order to prevent unaware breaking changes from upsream.

**Comment by iblis17** [comment]
@pluskid please checkout  8c5b6cf 

```julia
julia> mx._sig_checker()                        
WARNING: BatchNorm(data, gamma, beta, moving_mean, moving_var, eps, momentum, fix_gamma, use_global_stats, output_mean_var, axis, cudnn_off)
WARNING: CuDNNBatchNorm(data, eps, momentum, fix_gamma, use_global_stats, output_mean_var, axis, cudnn_off)
WARNING: Deconvolution(data, weight, bias, kernel, stride, dilate, pad, adj, target_shape, num_filter, num_group, workspace, no_bias, cudnn_tune, c
udnn_off, layout)                                                      
WARNING: GridGenerator(data, transform_type, target_shape)
WARNING: SliceChannel(data, num_outputs, axis, squeeze_axis)
WARNING: Softmax(data, grad_scale, ignore_label, multi_output, use_ignore, preserve_shape, normalization, out_grad, smooth_alpha)
WARNING: SoftmaxOutput(data, label, grad_scale, ignore_label, multi_output, use_ignore, preserve_shape, normalization, out_grad, smooth_alpha)
WARNING: SpatialTransformer(data, loc, target_shape, transform_type, sampler_type)
WARNING: _backward_slice_axis()                      
WARNING: _ones(shape, ctx, dtype)                                          
WARNING: _random_exponential(lam, shape, ctx, dtype)      
WARNING: _random_gamma(alpha, beta, shape, ctx, dtype)
WARNING: _random_generalized_negative_binomial(mu, alpha, shape, ctx, dtype)
WARNING: _random_negative_binomial(k, p, shape, ctx, dtype)
WARNING: _random_normal(loc, scale, shape, ctx, dtype)
WARNING: _random_poisson(lam, shape, ctx, dtype)
WARNING: _random_uniform(low, high, shape, ctx, dtype)
WARNING: _sample_exponential(lam, shape, dtype) 
WARNING: _sample_gamma(alpha, shape, dtype, beta)
WARNING: _sample_generalized_negative_binomial(mu, shape, dtype, alpha)
WARNING: _sample_multinomial(data, shape, get_prob, dtype)
WARNING: _sample_negative_binomial(k, shape, dtype, p)
WARNING: _sample_normal(mu, shape, dtype, sigma)
WARNING: _sample_poisson(lam, shape, dtype)
WARNING: _sample_uniform(low, shape, dtype, high)
WARNING: _sparse_sum(data, axis, keepdims, exclude)
WARNING: _square_sum(data, axis, keepdims, exclude)
WARNING: _zeros(shape, ctx, dtype)
WARNING: argmax(data, axis, keepdims)
WARNING: argmin(data, axis, keepdims)
WARNING: argsort(data, axis, is_ascend)
WARNING: broadcast_axes(data, axis, size)
WARNING: broadcast_axis(data, axis, size)
WARNING: broadcast_to(data, shape)
WARNING: expand_dims(data, axis)
WARNING: flip(data, axis)
WARNING: log_softmax(data, axis)
WARNING: nanprod(data, axis, keepdims, exclude)
WARNING: nansum(data, axis, keepdims, exclude)
WARNING: normal(loc, scale, shape, ctx, dtype)
WARNING: pick(data, index, axis, keepdims)
WARNING: random_exponential(lam, shape, ctx, dtype)
WARNING: random_gamma(alpha, beta, shape, ctx, dtype)
WARNING: random_generalized_negative_binomial(mu, alpha, shape, ctx, dtype)
WARNING: random_negative_binomial(k, p, shape, ctx, dtype)
WARNING: random_normal(loc, scale, shape, ctx, dtype)
WARNING: random_poisson(lam, shape, ctx, dtype)
WARNING: random_uniform(low, high, shape, ctx, dtype)
WARNING: repeat(data, repeats, axis)
WARNING: reshape_like(lhs, rhs)
WARNING: reverse(data, axis)
WARNING: sample_exponential(lam, shape, dtype)
WARNING: sample_gamma(alpha, shape, dtype, beta)
WARNING: sample_generalized_negative_binomial(mu, shape, dtype, alpha)
WARNING: sample_multinomial(data, shape, get_prob, dtype)
WARNING: sample_negative_binomial(k, shape, dtype, p)
WARNING: sample_normal(mu, shape, dtype, sigma)
WARNING: sample_poisson(lam, shape, dtype)
WARNING: sample_uniform(low, shape, dtype, high)
WARNING: scatter_nd(data, indices, shape)
WARNING: slice_axis(data, axis, begin, end)
WARNING: softmax(data, axis)
WARNING: sort(data, axis, is_ascend)
WARNING: split(data, num_outputs, axis, squeeze_axis)
WARNING: stack(data, axis, num_args)
WARNING: sum_axis(data, axis, keepdims, exclude)
WARNING: swapaxes(data, dim1, dim2)
WARNING: take(a, indices, axis, mode)
WARNING: topk(data, axis, k, ret_typ, is_ascend)
WARNING: uniform(low, high, shape, ctx, dtype)
```

**Comment by iblis17** [comment]
I can keep refining those APIs, but at this moment, I want to make a new release first. So I won't add more `remap` to this PR.

File: src/util.jl [hunk 1]
```diff
@@ -163,3 +163,28 @@ function _format_signature(narg::Int, arg_names::Ref{char_pp})
   return join([unsafe_string(name) for name in arg_names] , ", ")
 end
 
+@inline function _only2d(x)
+  @assert ndims(x) == 2
+  x
+end
+
+"""
+libmxnet operators signature checker.
```
Comment (pluskid) [inline src/util.jl]: Can you add more desc to the docstring?

**Comment by pluskid** [comment]
Thanks! Merged. But can you add a more detailed docstring for `_sig_checker`?

**Comment by iblis17** [comment]
sure.

**Comment by iblis17** [comment]
It's #305

**Pull request merged**